pkg/client/client_test.go

```diff
@@ -0,0 +1,236 @@
+
+func TestSuccessfulQuery(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	mqn := newMockQueryNet(testCid1, expectedQueryResponse)
+	require.NoError(t, mqn.setup())
+	defer func() {
+		require.NoError(t, mqn.teardown())
+	}()
+
+	go func() {
+		for ctx.Err() == nil {
+			select {
+			case <-ctx.Done():
+				return
+			case err := <-mqn.errChan:
+				cancel()
+				require.Fail(t, err.Error())
+				return
+			}
+		}
+	}()
+
+	ds1 := dss.MutexWrap(datastore.NewMapDatastore())
+	client, err := client.NewClient(ds1, mqn.client, nil)
+	require.NoError(t, err)
+
+	var connected bool
+	qr, err := client.RetrievalQueryToPeer(ctx, mqn.serverPeer(), testCid1, func() {
+		connected = true
+	})
+	require.NoError(t, err)
+	require.NotNil(t, qr)
+	require.True(t, connected)
+	require.Equal(t, &expectedQueryResponse, qr)
+}
+
+func TestBadQuery(t *testing.T) {
+	// in this test we're simulating an error on the query server side by making
+	// the mock query server expect a different CID, in which case it'll return
+	// an error back on its channel but mess up the query response
+	// we're also testing proper handling of context cancellation here too
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	mqn := newMockQueryNet(testCid1, expectedQueryResponse)
+	require.NoError(t, mqn.setup())
+	defer func() {
+		require.NoError(t, mqn.teardown())
+	}()
+
+	go func() {
+		for ctx.Err() == nil {
+			select {
+			case <-ctx.Done():
+				return
+			case err := <-mqn.errChan:
+				cancel()
+				require.Errorf(t, err, fmt.Sprintf("expected PayloadCID to be %s, got %s", testCid1, testCid2))
+				return
+			}
+		}
+	}()
+
+	ds1 := dss.MutexWrap(datastore.NewMapDatastore())
+	client, err := client.NewClient(ds1, mqn.client, nil)
+	require.NoError(t, err)
+
+	var connected bool
+	qr, err := client.RetrievalQueryToPeer(ctx, mqn.serverPeer(), testCid2, func() {
+		connected = true
+	})
+	require.ErrorIs(t, err, context.Canceled)
+	require.Nil(t, qr)
+	require.True(t, connected)
+	require.Nil(t, qr)
+}
+
+type mockQueryNet struct {
+	cid     cid.Cid
+	qr      retrievalmarket.QueryResponse
+	errChan chan error
+	mn      mocknet.Mocknet
+	server  host.Host
+	client  host.Host
+}
+
+func newMockQueryNet(cid cid.Cid, qr retrievalmarket.QueryResponse) *mockQueryNet {
+	mqn := &mockQueryNet{
+		cid:     cid,
+		qr:      qr,
+		errChan: make(chan error),
+	}
+	return mqn
+}
+
+func (mqn *mockQueryNet) serverPeer() peer.AddrInfo {
+	return peer.AddrInfo{ID: mqn.server.ID(), Addrs: mqn.server.Addrs()}
+}
+
+// Setup
+func (mqn *mockQueryNet) setup() error {
+	// setup network
+	var err error
+
+	mqn.mn = mocknet.New()
+	if mqn.server, err = mqn.mn.GenPeer(); err != nil {
+		return err
+	}
+	if mqn.client, err = mqn.mn.GenPeer(); err != nil {
+		return err
+	}
+	if err = mqn.mn.LinkAll(); err != nil {
+		return err
+	}
+
+	mqn.server.SetStreamHandler(retrievalmarket.QueryProtocolID, func(s network.Stream) {
```

Comment:
the complexity of this simple response points to a challenging long term issue when we build these -- I don't see how we respond to actual graphsync requests with filecoin retrieval negotiation happening on top without putting actual graphsync and the go-fil-markets retrieval server on top. Which don't get me wrong -- we can do, I just want to identify how hard that might end up being.